src/synthesizer.ts

```diff
@@ -0,0 +1,408 @@
+     * is required to be part of the stack name and construct ID to
+     * ensure uniqueness.
+     *
+     * @default 'staging-stack'
+     */
+    readonly stagingStackNamePrefix?: string;
+
+    /**
+     * Auto deletes objects in the staging S3 bucket and images in the
+     * staging ECR repositories.
+     *
+     * This will also delete the S3 buckets and ECR repositories themselves when
+     * all objects / images are removed.
+     *
+     * @default true
+     */
+    readonly autoDeleteStagingAssets?: boolean;
+}
+
+export class PulumiSynthesizer extends cdk.StackSynthesizer implements cdk.IReusableStackSynthesizer {
+    /**
+     * The Pulumi ComponentResource wrapper which contains all of the
+     * staging resources. This can be added to the `dependsOn` of the main
+     * stack to ensure the staging assets are created first
+     */
+    public readonly stagingStack: CdkConstruct;
+
+    /**
+     * The app-scoped, environment-keyed staging bucket.
+     */
+    public stagingBucket?: aws.s3.BucketV2;
+
+    /**
+     * The app-scoped, environment-keyed ecr repositories associated with this app.
+     */
+    public readonly stagingRepos: Record<string, aws.ecr.Repository> = {};
+
+    private readonly appId: string;
+
+    /**
+     * A value to use as an override for the bucket name
+     */
+    private readonly _stagingBucketName?: string;
+    private readonly autoDeleteStagingAssets: boolean;
+
+    /**
+     * The final CDK name of the asset S3 bucket. This may contain CDK tokens
+     */
+    private cdkBucketName?: string;
+
+    /**
+     * The final Pulumi name of the asset S3 bucket
+     */
+    private pulumiBucketName?: string | pulumi.Output<string>;
+
+    /**
+     * The region from the pulumi provider
+     */
+    private readonly pulumiRegion: pulumi.Output<string>;
+
+    /**
+     * The account id from the pulumi provider
+     */
+    private readonly pulumiAccount: pulumi.Output<string>;
+
+    /**
+     * The accountId which may contain CDK tokens
+     */
+    private cdkAccount?: string;
+
+    /**
+     * The region that may contain CDK tokens
+     */
+    private cdkRegion?: string;
+
+    /**
+     * The resources that any file assets need to depend on
+     */
+    private readonly fileDependencies: pulumi.Resource[] = [];
+
+    private readonly assetManifest = new cdk.AssetManifestBuilder();
+
+    /**
+     * The output directory which contains the asset files.
+     * Can be used to generate the absolute path to the asset
+     */
+    private outdir?: string;
+
+    constructor(props: PulumiSynthesizerOptions) {
+        super();
+        const stackPrefix = props.stagingStackNamePrefix ?? 'staging-stack';
+        this._stagingBucketName = props.stagingBucketName;
+        this.autoDeleteStagingAssets = props.autoDeleteStagingAssets ?? true;
+        this.appId = this.validateAppId(props.appId);
+
+        const caller = aws.getCallerIdentityOutput();
```

Comment:
Based on our conversation I'm leaning towards deferring this until https://github.com/pulumi/pulumi-cdk/pull/170 since we can then just inherit the provider from the parent.

Reply:
Sounds good! Can you please add a todo in the code so we don't forget about it?